src/core/lib/event_engine/grpc_ares_wrapper.h

```diff
@@ -121,95 +141,64 @@ class GrpcAresRequest
   // close the socket (possibly through ares_destroy).
   struct FdNode {
     FdNode() = default;
-    FdNode(ares_socket_t as, GrpcPolledFd* polled_fd);
+    FdNode(ares_socket_t as, GrpcPolledFd* polled_fd)
+        : as(as), polled_fd(polled_fd) {}
     ares_socket_t as;
     std::unique_ptr<GrpcPolledFd> polled_fd;
-    // next fd node
-    FdNode* next = nullptr;
     // true if the readable closure has been registered
     bool readable_registered = false;
     // true if the writable closure has been registered
     bool writable_registered = false;
     bool already_shutdown = false;
   };
+  using FdNodeList = std::list<std::unique_ptr<FdNode>>;
 
-  // A linked-list of FdNodes. Support operations such as pop a FdNode which has
-  // a specific socket/fd that c-ares owns.
-  class FdNodeList {
-   public:
-    class FdNodeListIterator {
-     public:
-      bool operator!=(const FdNodeListIterator& other) {
-        return node_ != other.node_;
-      }
-      bool operator==(const FdNodeListIterator& other) {
-        return node_ == other.node_;
-      }
-      FdNodeListIterator& operator++(int) {
-        node_ = node_->next;
-        return *this;
-      }
-      FdNode* operator*() { return node_; }
-      static FdNodeListIterator universal_end() {
-        return FdNodeListIterator(nullptr);
-      }
-
-     private:
-      friend class FdNodeList;
-      explicit FdNodeListIterator(FdNode* node) : node_(node) {}
-      FdNode* node_;
-    };
-    using Iterator = FdNodeListIterator;
-
-    ~FdNodeList() { GPR_ASSERT(IsEmpty()); }
-
-    Iterator begin() { return Iterator(head_); }
-    Iterator end() { return Iterator::universal_end(); }
-
-    bool IsEmpty() const { return head_ == nullptr; }
-
-    void PushFdNode(FdNode* fd_node) ABSL_EXCLUSIVE_LOCKS_REQUIRED(&mu_) {
-      fd_node->next = head_;
-      head_ = fd_node;
-    }
-
-    FdNode* PopFdNode() ABSL_EXCLUSIVE_LOCKS_REQUIRED(&mu_) {
-      GPR_ASSERT(!IsEmpty());
-      FdNode* ret = head_;
-      head_ = head_->next;
-      return ret;
-    }
-
-    // Search for as in the FdNode list. This is an O(n) search, the max
-    // possible value of n is ARES_GETSOCK_MAXNUM (16). n is typically 1 - 2
-    // in our tests.
-    FdNode* PopFdNode(ares_socket_t as) ABSL_EXCLUSIVE_LOCKS_REQUIRED(&mu_);
-
-   private:
-    FdNode* head_ = nullptr;
-  };
   friend class GrpcPolledFdFactory;
 
+  // Requires mu_ to be locked before calling.
+  // Unlock mu_ before returns.
+  //
+  // Ideally this could take a lock object by reference or something like a
+  // std::unique_ptr<grpc_core::ReleasableMutexLock>. But Thread Safety Analysis
+  // does not work in these scenarios, see for example:
+  // https://github.com/llvm/llvm-project/issues/59662.
+  virtual void OnResolveLockedAndUnlock() ABSL_UNLOCK_FUNCTION(mu()) = 0;
```

Comment:
In the new implementation, this function is not needed anymore. We will rely on `event_engine->Run`   to run the caller's callback which *must not run it in the calling thread before Run returns*. This would prevent us from having to acquire locks in reverse order. (responding ahead of time since the next push will make this comment outdated)